pkg/csi/hostpathimpl.go

```diff
@@ -0,0 +1,264 @@
+		return nil, status.Error(codes.Unimplemented, "CreateVolume: volume content source is not supported")
+	}
+
+	reqCtx := getValue(ctx, ctxKeyCreateVolume)
+	nodeName := reqCtx.nodeName
+	volumeID := req.GetName()
+	parameters := req.GetParameters()
+
+	// TODO(x.zhou): interact with the scheduler to see if there is sufficient space to allocate
+
+	parameters[pkg.AnnoSelectedNode] = nodeName
+	parameters[volumeCapacityTag] = fmt.Sprintf("%d", req.GetCapacityRange().GetRequiredBytes())
+	response := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId:      volumeID,
+			CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
+			VolumeContext: parameters,
+			AccessibleTopology: []*csi.Topology{{
+				Segments: map[string]string{
+					pkg.KubernetesNodeIdentityKey: nodeName,
+				},
+			}},
+		},
+	}
+	log.Infof("CreateVolume: create volume %s(size: %d) successfully", volumeID, req.GetCapacityRange().GetRequiredBytes())
+	return response, nil
+}
+
+func (impl *hostPathImpl) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	reqCtx := getValue(ctx, ctxKeyDeleteVolume)
+	volumeID := reqCtx.pv.Name
+	parameters := reqCtx.pv.Spec.CSI.VolumeAttributes
+	_, hostPath, err := impl.getVerifiedHostPath(volumeID, parameters)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "DeleteVolume: getVerifiedHostPath error: %s", err.Error())
+	}
+	nodeName := reqCtx.nodeName
+
+	conn, err := impl.cs.getNodeConn(nodeName)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "DeleteVolume: failed to connect to node %s: %s", nodeName, err.Error())
+	}
+	defer conn.Close()
+	if err := conn.CleanPath(ctx, hostPath, true); err != nil {
+		return nil, status.Errorf(codes.Internal, "DeleteVolume: failed to delete hostPath %s: %s", hostPath, err.Error())
+	}
+	log.Infof("DeleteVolume: delete HostPath volume(%s) successfully", volumeID)
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+func (impl *hostPathImpl) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (resp *csi.NodePublishVolumeResponse, err error) {
+	volumeID := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
+
+	basePath, volumeHostPath, err := impl.getVerifiedHostPath(volumeID, req.VolumeContext)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "NodePublishVolume: getVerifiedHostPath error: %s", err.Error())
+	}
+
+	// create volume directory on the host
+	if err := impl.ns.osTool.MkdirAll(volumeHostPath, 0777); err != nil {
+		return nil, status.Errorf(codes.Internal, "NodePublishVolume: failed to create directory %s: %s", volumeHostPath, err.Error())
+	}
+	defer func() {
+		// clean up volume directory when error
+		if err != nil {
+			_ = impl.ns.osTool.Remove(volumeHostPath)
+		}
+	}()
+
+	// set project quota
+	setQuotaDone := false
+	var setQuotaErr error
+	tags := tags(req.VolumeContext)
+	capacity, err := tags.GetInt64(volumeCapacityTag)
```

Comment:
better to check if capacity is less than zero

Reply:
Done.